smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -829,6 +1001,201 @@ private void readHeaders(
+                });
+                Symbol targetSymbol = context.getSymbolProvider().toSymbol(target);
+                StringBuilder deserFunctionBuilder = new StringBuilder(ProtocolGenerator.getDeserFunctionName(
+                        targetSymbol, context.getProtocolName())).append("_event");
+                if (target instanceof StructureShape) {
+                    // Single-event stream. Save the structure and generate event-specific deser later.
+                    this.deserializingEventShapes.add(target.asStructureShape().get());
+                    // For single-event stream, supply event message to corresponding event structure deser.
+                    deserFunctionBuilder.append("(eventMessage, context)");
+                } else if (target instanceof UnionShape) {
+                    // Multi-event stream. Save the union and generate dispatcher later.
+                    this.deserializeEventUnions.add(target.asUnionShape().get());
+                    // For multi-event stream, supply event name to event pairs to the events union deser.
+                    deserFunctionBuilder.append("(parsedEvent, context)");
+                } else {
+                    throw new CodegenException(String.format("Unexpected shape targeted by eventstream: `%s`",
+                            target.getType()));
+                }
+                writer.write("return await $L;", deserFunctionBuilder.toString());
+
+            });
+        });
+    }
+
+    /**
+     * Writes a function used to dispatch event to corresponding event deserializers if given
+     * event stream is a multi-event event stream.
+     */
+    private void generateDeserializingEventUnion(
+            GenerationContext context,
+            UnionShape events
+    ) {
+        TypeScriptWriter writer = context.getWriter();
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        Symbol symbol = symbolProvider.toSymbol(events);
+        String protocolName = context.getProtocolName();
+        String methodName = ProtocolGenerator.getDeserFunctionName(symbol, protocolName) + "_event";
+        Model model = context.getModel();
+        writer.openBlock("const $L = async (\n"
+                       + "  output: any,\n"
+                       + "  context: __SerdeContext\n"
+                       + "): Promise<$T> => {", "}", methodName, symbol, () -> {
+            events.getAllMembers().forEach((name, member) -> {
+                StructureShape target = model.expectShape(member.getTarget(), StructureShape.class);
+                // Prepare event for generating event deserializers.
+                deserializingEventShapes.add(target);
+                writer.openBlock("if (output[$S] !== undefined) {", "}", name, () -> {
+                    writer.openBlock("return {", "};", () -> {
+                        // Dispatch to special event deserialize function
+                        Symbol eventSymbol = symbolProvider.toSymbol(target);
```

Comment:
This section can be updated for cleanliness, using positional arguments and the different formatters:

```java
Symbol eventSymbol = symbolProvider.toSymbol(target);
String methodName = ProtocolGenerator.getDeserFunctionName(eventSymbol, protocolName) + "_event";
writer.write("$1L: await $2L(output[$1S], context)", name, methodName)
```